file-attribute-caching/src/main/kotlin/com/pkware/file/fileAttributeCaching/FileAttributeCachingFileSystemProvider.kt

```diff
@@ -0,0 +1,352 @@
+package com.pkware.file.fileAttributeCaching
+
+import com.google.auto.service.AutoService
+import java.io.IOException
+import java.net.URI
+import java.nio.channels.SeekableByteChannel
+import java.nio.file.AccessMode
+import java.nio.file.CopyOption
+import java.nio.file.DirectoryStream
+import java.nio.file.FileStore
+import java.nio.file.FileSystem
+import java.nio.file.FileSystemNotFoundException
+import java.nio.file.Files
+import java.nio.file.LinkOption
+import java.nio.file.OpenOption
+import java.nio.file.Path
+import java.nio.file.Paths
+import java.nio.file.StandardCopyOption
+import java.nio.file.attribute.BasicFileAttributeView
+import java.nio.file.attribute.BasicFileAttributes
+import java.nio.file.attribute.DosFileAttributeView
+import java.nio.file.attribute.DosFileAttributes
+import java.nio.file.attribute.FileAttribute
+import java.nio.file.attribute.FileAttributeView
+import java.nio.file.attribute.PosixFileAttributeView
+import java.nio.file.attribute.PosixFileAttributes
+import java.nio.file.spi.FileSystemProvider
+import java.util.Locale
+
+/**
+ * A [FileSystemProvider] wrapper that handles [FileAttributeCachingPath]s for reading and writing file attributes.
+ *
+ * It forwards most operations using the [FileAttributeCachingPath]s underlying delegate [Path]. It caches file
+ * attributes ([BasicFileAttributes], [DosFileAttributes], and [PosixFileAttributes]) on reads, writes, copy and move
+ * operations.
+ */
+@AutoService(FileSystemProvider::class)
+internal class FileAttributeCachingFileSystemProvider : FileSystemProvider() {
```

Comment:
It was a design decision to have the cache exist per path rather than in the filesystemProvider or anywhere else. 

To keep track of all the path(s) and their caches in the provider would be a nightmare. You also cannot have one cache for all of the path(s) unless you also want to keep track of each path(s) expiration time

Reply:
> I just don't know if this is fixable with our current approach or not

I'm assuming it's fixable even with the current approach, but not without becoming unnecessarily complicated and harder to follow.

> It seems dangerous from an outside perspective to assume that all paths they would pass to us will be the type of paths we would expect.

I would actually go one step further and say that, more importantly than the danger, this is about ease of use. In Bayer's core-libraries branch, the only changes right now are using the fileSystem, and then we get everything for free. That's what I would expect if we're making a change at the fileSystem level, caching "just works" for anything involving paths. That's what we should strive for. It seems like a more realistic possibility to achieve that at the provider level.

Reply:
So what else should I be covering here? that we always return cachingPath's for every operation that involves a path? 

I specifically DONT assume that any path coming in is a caching one and cast it to such in the provider functions- thus we would get a casting exception if it cannot be converted. If i assumed every path would be a caching one then I wouldnt need the checks tbh.

Also, this should have been brought up WAYYYY earlier in the design process if this was actually a concern considering Ive been working on this since early november.

Reply:
Also I have no idea how it would work to just have the provider do all of the caching and access it within core libraries- at that point I would say my original shim solution is a better one

Reply:
> Also, this should have been brought up WAYYYY earlier in the design process if this was actually a concern considering Ive been working on this since early november.

I did originally question the cache happening at the provider level and not the path level when you and I had a call, but I digress.

We're here today. And today, what Ryan is saying that, regardless of what our solution is going forward, this solution falls short of properly addressing the IOPS issue. That doesn't mean that you, @mbayerPK, have done anything wrong. That means that we, the team responsible for the solution, are criticizing the solution we have arrived at as a team and are looking to address that as a team.

He has presented two options.

1. To handle caching at the provider level, featuring an expiring cache (of Path -> APathAndTheStateOfItsAttributes) for every path we encounter. This is reasonable.
2. To eliminate the expiration of paths and make sure this approach is more resilient to the way 3rd party libraries might process paths and invoke path-related operations. While these two components aren't mutually exclusive, I will say that the original design discussed with Marius made _**no**_ mention of expiring at the path level, instead favoring the idea of allowing garbage collection to handle cleanup of something like our `FileAttributeCachingPath`.

The important thing to note about these 2 options is they solve entirely different problems.

- If our goal is to cache _**everything**_ coming through the filesystem, then option 1 is superior.
- If our goal is to cache only the items on a specific path, then option 2 is definitely the way to go.

But if option 2 is the solution (and I think it might be),  then we need to do improve the resiliency of paths and the transitions they encounter. I'd recommend working with debugging and modifying an `XlsxCrawler` flow to use one of your `FileAttributeCachingPath` objects and paying close attention to if your attribute methods are hit. As Ryan alluded to,  I'm assuming they will not. Let's fix that.

Reply:
the goal was never to cache everything, the goal was to cache only the file/path currently being scanned/supplied to the unpacker. 

In terms of an unpacker workflow that would be only the original file, and we only care about the attributes before a given operation and restoring them after the operation completes- we dont really need cached attributes for every intermediate temp file / path created by workflows in the meantime.

Reply:
> the goal was never to cache everything, the goal was to cache only the file/path currently being scanned/supplied to the unpacker.
> 
> In terms of an unpacker workflow that would be only the original file, and we only care about the attributes before a given operation and restoring them after the operation completes- we dont really need cached attributes for every intermediate temp file / path created by workflows in the meantime.

But the meta-goal we're striving for is for reducing intermediate IOPS. And if the intermediate steps nullify our approach (either by prematurely expiring cache entries, or operating on objects that don't have attributes cached), we won't accomplish that.

Reply:
the goal with this was to replace my shim that i made with the unpacker context and keep iops at least the same OR better them

Reply:
> keep iops at least the same

This will be worse for any scan lasting longer than 5 seconds.

> OR better them

This is unlikely to happen if the path is altered in anyway, like in most of our crawlers where we call `.toFile()`, or others, where the 3rd party libraries that _do_ accept paths call things like `toAbsolutePath` and vacate the original object, and thus incur IOPS.

Reply:
we dont have to incur extra iops if the cache is copied for the various operations we care about along the way.

I feel like im being told all of the work ive done is pointless and im just wasting my time at this point

Reply:
> we dont have to incur extra iops if the cache is copied for the various operations we care about along the way.

Exactly why we should probably override functions like `toAbsolutePath` to return a `FileAttributeCachingPath` with a copy of our cache, its delegate being the original delegate's `toAbsolutePath`.

> I feel like im being told all of the work ive done is pointless and im just wasting my time at this point

Don't ever get attached to your software babies. Take as good of care as possible, but they all eventually die. Just hope you helped them succeed enough to give birth to something better.

Reply:
okay so we are not adopting the filesystem provider based approach for now but implementing other changes

@PKJosh will help with getting some simple procmon based performance tests going with this product's integration in core libraries